engine/execution/ingestion/engine.go

```diff
@@ -243,12 +243,15 @@ func (e *Engine) finalizedUnexecutedBlocks(finalized protocol.Snapshot) ([]flow.
 		}
 
 		unexecuted = append(unexecuted, header.ID())
+		if len(unexecuted) >= maxBlocks {
+			return unexecuted, nil
+		}
 	}
 
 	return unexecuted, nil
 }
 
-func (e *Engine) pendingUnexecutedBlocks(finalized protocol.Snapshot) ([]flow.Identifier, error) {
+func (e *Engine) pendingUnexecutedBlocks(finalized protocol.Snapshot, maxBlocks int) ([]flow.Identifier, error) {
 	pendings, err := finalized.ValidDescendants()
```

Comment:
The most common case is there are too many unexecuted and finalized blocks, so maxBlocks here would be `0`. And we could `ValidDescendants`. 
```suggestion
	if maxBlocks == 0 {
		return []flow.Identifier{}, nil  
	}
	pendings, err := finalized.ValidDescendants()
```